src/charmed_kubeflow_chisme/testing/_unit_tests.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""
+# Testing
+
+Tools for unit or integration testing, such as importable and reusable tests.
+
+## Usage example
+
+```python
+import pytest
+from charm import Operator
+from charmed_kubeflow_chisme.testing import test_leadership_events as leadership_events
+from charmed_kubeflow_chisme.testing import test_missing_relation as missing_relation
+from ops.model import WaitingStatus
+from ops.testing import Harness
+
+
+@pytest.fixture
+def harness():
+    return Harness(Operator)
+
+
+def test_leadership_events(harness):
+    leadership_events(harness)
+
+
+def test_missing_relation(harness):
+    missing_relation(harness, WaitingStatus, oci_image_added=False)
+```
+"""
+
+from contextlib import nullcontext as does_not_raise
+
+import pytest
+from oci_image import MissingResourceError
+from ops.model import BlockedStatus, WaitingStatus
+
+
+def test_leadership_events(harness):
+    """Test leader-elected event handling.
+
+    Tests if the unit raises correct status:
+
+    * `WaitingStatus` when it's not a leader
+
+    * when elected as the leader, it should leave `WaitingStatus`
+
+    * `WaitingStatus` when another unit is elected as the leader.
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+    """
+    harness.set_leader(False)
+    harness.begin_with_initial_hooks()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    harness.set_leader(True)
+    assert harness.charm.model.unit.status != WaitingStatus("Waiting for leadership")
+    harness.set_leader(False)
+    # Emit another leader_elected event due to https://github.com/canonical/operator/issues/812
+    harness._charm.on.leader_elected.emit()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+
+def test_missing_image(harness, expected_status=BlockedStatus, leader_check=True):
+    """Tests if the unit raises an expected status when a required oci image is missing
+    in a charm with the following checks order:
+
+    1) check for leadership (optional)
+
+    2) check oci image
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+        expected_status: a subclass of `ops.model.StatusBase`. Default: `BlockedStatus`
+        leader_check: whether the unit should be set to leader first. Default: True
+    """
+    if leader_check:
+        harness.set_leader(True)
+    harness.begin_with_initial_hooks()
+    assert isinstance(harness.charm.model.unit.status, expected_status)
```

Comment:
Given that you have optional arguments, what we essentially have is a test case that wraps the following calls:

```python
    harness.begin_with_initial_hooks()
    assert isinstance(harness.charm.model.unit.status, expected_status)
```

I could use these test cases for testing things they are not intended to, for example, I could use them in the `training-operator` (a sidecar pattern charm) that has zero calls to the checks you are expecting here (like `_check_for_image_details`) for just asserting the status of that charm after initial hooks. I could even use both `test_missing_image` and `test_missing_relation`, interchangeably.
That being said, I cannot see how in their current state, they add value.

This same comment applies to `test_missing_relation`.

Reply:
I get your point. Those tests are based on podspec charms, which follow a particular order of checks. For example, most kfp charms could use these tests in the current state, as shown in [tensorboards draft](https://github.com/canonical/kubeflow-tensorboards-operator/pull/25).
In that case shall we just keep `leadership_events` and `not_kubeflow_model` to have the shared testing structure or consider closing this PR and adding new tests once we rewrite more charms to sidecar and identify common patterns? @dnplas @ca-scribner thoughts?

Reply:
> In that case shall we just keep leadership_events and not_kubeflow_model to have the shared testing structure or consider closing this PR and adding new tests once we rewrite more charms to sidecar and identify common patterns?

IMHO, the second option will better suit our mid term goals. This is a great effort, and I believe all the team will appreciate shared tests, but we are moving away from podspec to a different pattern, which will make these tests obsolete very soon.